prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,537 @@
+	}
+
+	return page_count;
+}
+
+static size_t __mbox_count(struct gnix_mbox_alloc_handle *handle)
+{
+	size_t mbox_count = (__page_count(handle) * handle->page_size) /
+			    handle->mbox_size;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Mbox_count: %zu.\n", mbox_count);
+	return mbox_count;
+}
+
+static int __create_slab(struct gnix_mbox_alloc_handle *handle)
+{
+	struct gnix_slab *slab;
+	gni_return_t status;
+	size_t total_size;
+	int ret;
+
+	slab = calloc(1, sizeof(*slab));
+	if (!slab) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating slab: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "total_size requested for mmap: %zu.\n",
+		   total_size);
+
+	slab->used = calloc(1, sizeof(*(slab->used)));
+	if (!slab->used) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating bitmap: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	slab->base = mmap(0, total_size, (PROT_READ | PROT_WRITE), MAP_SHARED,
+			  handle->fd, handle->last_offset);
+	if (slab->base == MAP_FAILED) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "%s\n", strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err2;
+	}
+
+	ret = alloc_bitmap(slab->used, __mbox_count(handle));
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error allocating bitmap.\n");
+		goto err3;
+	}
+
+	status = GNI_MemRegister(handle->nic_handle->gni_nic_hndl,
+				 (uint64_t) slab->base, total_size,
+				 handle->cq_handle,
+				 GNI_MEM_READWRITE, -1,
+				 &slab->memory_handle);
+	if (status != GNI_RC_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "GNI_MemRegister failed: %s\n",
+			  gni_err_str[status]);
+		ret = gnixu_to_fi_errno(status);
+		goto err4;
+	}
+
+	slab->allocator = handle;
+
+	slist_insert_tail(&slab->list_entry, &handle->slab_list);
+
+	handle->last_offset += total_size;
+
+	return ret;
+
+err4:
+	free_bitmap(slab->used);
+	free(slab->used);
+err3:
+	munmap(slab->base, total_size);
+err2:
+	free(slab->used);
+err1:
+	free(slab);
+err:
+	return ret;
+}
+
+/*
+ * TODO:
+ * - How to handle memory that is still out...
+ */
+static int __destroy_slab(struct gnix_mbox_alloc_handle *handle,
+			  struct gnix_slab *slab)
+{
+	size_t total_size;
+
+	if (!handle || !slab) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid argument handle or slab.\n");
+		return -FI_EINVAL;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+
+	free_bitmap(slab->used);
+	free(slab->used);
+
+	GNI_MemDeregister(handle->nic_handle->gni_nic_hndl,
+			  &slab->memory_handle);
+
+	munmap(slab->base, total_size);
+
+	free(slab);
+
+	return FI_SUCCESS;
+}
+
+static int __find_free(struct gnix_mbox_alloc_handle *handle,
+		       struct gnix_slab **slab)
+{
+	struct slist_entry *entry;
+	struct gnix_slab *temp;
+	int ret;
+
+	for (entry = handle->slab_list.head; entry; entry = entry->next) {
+		temp = container_of(entry, struct gnix_slab, list_entry);
+		ret = find_first_zero_bit(temp->used);
+		if (ret >= 0)
```

Comment:
Instead of doing the same comparison twice, why not do the following?
```C
// ...
*slab = NULL;

// for ....
        if (ret >= 0) {
            *slab = temp;
            break;
        } 
```

* If there are no allocated slabs, slab won't be set to an uninitialized pointer (is this possible with your implementation?)
* cleaner in terms of the instruction pipeline